### PR TITLE
The ability to launch up to four instances from the Editor on Build.

### LINKED ISSUE #4075: Feature request: Single/multiple instance mode (DEFEDIT-1590)
As a developer, I want to be able to run multiple instances of the game from one editor instance. For example, when testing multiplayer, it can be very usefull to have more than one instances of the game running. Bundling it every time makes process really slow. It would be really good if you could specify if the editor should allow multiple instances of the engine running.

For example IntelliJ IDEA has this option.

### PR DESCRIPTION
**Pull request by AGulev**
Now, in the `Project` menu, it is possible to pick how many instances the Editor should launch when building. This functionality is useful for developing a multiplayer project. The following features are supported:

- Launch multiple instances of the engine.
- Relaunch or hot-reload all instances simultaneously.
- Relaunch, hot-reload, and debug each instance separately if a particular target is selected in the `Project -> Target` menu.
- If more than one instance is launched, the instance index will be added to the window title.
- The index is available at runtime using `local instance_index = sys.get_config_int("project.instance_index", 0)`.
- If logs are enabled, each instance will use its own log file, `instance_N_log.txt`, where N is the instance index.
- If more than one instance is launched, each instance will have its own debugger port: the default port + instance index.
- A new menu option (and shortcut `Ctrl+J` or 'Cmd+J' on Mac) has been added to close the engine (all instances) from the editor. See the `Project` menu.

CleanShot 2024-08-05 at 13 21 04@2x

Fix https://github.com/defold/defold/issues/4075
## PR checklist

* [x] Code
	* [ ] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [x] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [x] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [x] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [x] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [x] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: editor/src/clj/editor/app_view.clj
Comment (AGulev) [inline editor/src/clj/editor/app_view.clj:789]: If seems like it's not the main UI-thread. But even if it is, it shouldn't be a problem, right?

File: editor/src/clj/editor/app_view.clj
Comment (AGulev) [inline editor/src/clj/editor/app_view.clj:849]: Should I maybe pass `instance-index` into the hook as well?

File: editor/src/clj/editor/app_view.clj
Comment (AGulev) [inline editor/src/clj/editor/app_view.clj:853]: we get address and the other useful info from the first few lines in logs. So, we have to do it for each target we launch, but keep console connected only to the last target

File: editor/src/clj/editor/app_view.clj
Comment (AGulev) [inline editor/src/clj/editor/app_view.clj:1245]: Do I need atom here? I'm not sure because I saw both implementation with and without atom.

File: editor/src/clj/editor/app_view.clj
Comment (AGulev) [inline editor/src/clj/editor/app_view.clj:1408]: magic of hot-reloading for all the instances! 

https://github.com/user-attachments/assets/fc9024d0-7879-48fd-ac2c-fc6e67cd3b41

File: editor/src/clj/editor/debug_view.clj
Comment (AGulev) [inline editor/src/clj/editor/debug_view.clj:489]: Use a separate debugger port for each instance to be able to debug them all (not simultaneously, no, picking one in Target menu)

File: editor/src/clj/editor/engine.clj
Comment (AGulev) [inline editor/src/clj/editor/engine.clj:100]: notify the engine that this is one of a few instances: it used in `engine.cpp` and in debugger.lua

File: editor/src/clj/editor/engine.clj
Comment (britzl) [inline editor/src/clj/editor/engine.clj:273]: I wonder if maybe we should set the instance_index as a process environment variable instead of a project property? Like DM_SERVICE_PORT and DM_QUIT_ON_ESC? Perhaps DM_ENGINE_INSTANCE_INDEX?

At runtime it can probably be read using `os.getenv("DM_ENGINE_INSTANCE_INDEX")` or if we expose `dmSys::GetEnv()` to Lua?

File: editor/src/clj/editor/keymap.clj
Comment (AGulev) [inline editor/src/clj/editor/keymap.clj:101]: If you have better ideas for a shortcut, you are welcome. 
This shortcut is needed because if we don't add it - it will be annoying to close all the editors...

File: editor/src/clj/editor/prefs.clj
Comment (AGulev) [inline editor/src/clj/editor/prefs.clj:69]: it was moves from `recent_files.clj` to be able to reuse

File: editor/src/clj/editor/targets.clj
Comment (AGulev) [inline editor/src/clj/editor/targets.clj:230]: Names do not contain the port, so if we have a few instances the names for them will be the same, which might be a reason of infinite updating of the menu (I caught it many times)

**Review by britzl — COMMENTED** [review]
Looks good. But I have a question about the instance index

File: editor/src/clj/editor/targets.clj
Comment (AGulev) [inline editor/src/clj/editor/targets.clj:376]: I think this is a good name, but if you don't think so, gimme your options.

File: engine/engine/content/builtins/scripts/debugger.lua
Comment (AGulev) [inline engine/engine/content/builtins/scripts/debugger.lua:181]: use a separate prot for each instance

File: engine/engine/src/engine.cpp
Comment (AGulev) [inline engine/engine/src/engine.cpp:816]: custom log for each instance, but keep 0 as old behavior.

File: engine/engine/src/engine.cpp
Comment (AGulev) [inline engine/engine/src/engine.cpp:865]: add index into the title. It's convenient, but also it gives each windows position save on mac for free (it's OS functionality we use now in glfw3)

File: editor/src/clj/editor/engine.clj
Comment (AGulev) [reply editor/src/clj/editor/engine.clj:273]: Yes, we can. But is it somehow better?

File: editor/src/clj/editor/defold_project.clj
Comment (AGulev) [inline editor/src/clj/editor/defold_project.clj:668]: Naming. Again. If you have better options - let me know

File: editor/src/clj/editor/defold_project.clj
Comment (AGulev) [inline editor/src/clj/editor/defold_project.clj:663]: is it good enough? Should we change manu names somehow here?
CleanShot 2024-08-02 at 11 43 01@2x

File: editor/src/clj/editor/engine.clj
Comment (britzl) [reply editor/src/clj/editor/engine.clj:273]: I think it is. Entries in game.project are mostly user-defined right? Do we have other example of where we inject values into project settings on startup? In the case of the instance count it is some extra data which the editor defines and provides on startup, much like the service port, which is why I'm thinking that an env variable makes more sense.

Comment (AGulev) [reply editor/src/clj/editor/engine.clj:273]: > Do we have other example of where we inject values into project settings on startup?

yes, we use it in a few proaces: 
https://github.com/defold/defold/blob/8317584212f3047dd6881f38ba6f4416dcf16bb1/editor/src/clj/editor/engine.clj#L90

https://github.com/defold/defold/blob/8317584212f3047dd6881f38ba6f4416dcf16bb1/editor/src/clj/editor/engine.clj#L92

https://github.com/defold/defold/blob/8317584212f3047dd6881f38ba6f4416dcf16bb1/editor/src/clj/editor/engine.clj#L262

https://github.com/defold/defold/blob/8317584212f3047dd6881f38ba6f4416dcf16bb1/editor/src/clj/editor/engine.clj#L265


I don't have a strong opinion here. If we want - I can rewrite it. It will be a bit more Lua code to get default value 0, but it's not critical.

Comment (britzl) [reply editor/src/clj/editor/engine.clj:273]: Ah, ok, if there is a precedent, then let us keep it as it is now!

**Review by britzl — DISMISSED** [review]

File: editor/src/clj/editor/targets.clj
Comment (matgis) [reply editor/src/clj/editor/targets.clj:376]: I agree. It's good!

**Review by matgis — DISMISSED** [review]
Very nice feature addition! It works really well! I love that it remembers the window positions between launches, too!

I found some minor things and a few whitespace issues, but I'll let you decide if you want to address them before merging.

You are probably aware of this already, but during testing, I noticed the two **Open Profiler** commands under the **Debug** menu will only connect to one of the running instances (presumably the last one?). We might want to enable it to connect to a particular instance in a separate PR later. Perhaps we should open an issue to track it?

I also noticed that the **Simulate Resolution** function can't target all running instances at once, but again, we can address that later in a separate issue.

File: editor/src/clj/editor/defold_project.clj
Comment (AGulev) [reply editor/src/clj/editor/defold_project.clj]: Fixed

**Comment by AGulev** [comment]
@matgis thank you. 
I made all the fixes you requested. 

Yes, I didn't pay too much attention to the profiler, because it feels like not the best idea to profile something running many instances. But we will fix it if somebody request it for sure.

The resolution thing is one I forgot about. I'll fix it as part of this issue I'm going to take a look soon https://github.com/defold/defold/issues/8130

File: editor/src/clj/editor/defold_project.clj
Comment (AGulev) [reply editor/src/clj/editor/defold_project.clj]: Fixed

**Review by matgis — APPROVED** [review]
Ship it! ❤️

**Pull request merged**
